modules/mongodb/scanner.go

```diff
@@ -0,0 +1,346 @@
+	out := make([]byte, msglen)
+	// msg header
+	binary.LittleEndian.PutUint32(out[0:], uint32(msglen))
+	binary.LittleEndian.PutUint32(out[12:], OP_COMMAND)
+	// command msg
+	idx := MSGHEADER_LEN
+	copy(out[idx:idx+dblen], []byte(database))
+	idx += dblen
+	copy(out[idx:idx+cnlen], []byte(commandName))
+	idx += cnlen
+	copy(out[idx:idx+mdlen], metadata)
+	idx += mdlen
+	copy(out[idx:idx+calen], commandArgs)
+	return out
+}
+
+// getIsMasterMsg returns a mongodb message containing isMaster command.
+// https://docs.mongodb.com/manual/reference/command/isMaster/
+func getIsMasterMsg() ([]byte) {
+	query, err := bson.Marshal(bson.M{ "isMaster": 1 })
+	if err != nil {
+		// programmer error
+		log.Fatal("Invalid BSON")
+	}
+	query_msg := getOpQuery("admin.$cmd", query)
+	return query_msg
+}
+
+// getBuildInfoCommandMsg returns a mongodb message containing a command to retrieve MongoDB build info.
+func getBuildInfoCommandMsg() ([]byte) {
+	metaData, err := bson.Marshal(bson.M{ "buildInfo": 1 })
+	if err != nil {
+		// programmer error
+		log.Fatal("Invalid BSON")
+	}
+	commandArgs, err := bson.Marshal(bson.M{})
+	if err != nil {
+		// programmer error
+		log.Fatal("Invalid BSON")
+	}
+	// "test" collection gleaned from tshark
+	command_msg := getCommandMsg("test", "buildInfo", metaData, commandArgs)
+	return command_msg
+}
+
+// getOpQuery returns a mongodb OP_QUERY message containing the specified BSON-encoded query.
+// query expected to be BSON byte array.
+func getOpQuery(collname string, query []byte) ([]byte) {
+	flagslen := 4
+	collname_len := len(collname) + 1
+	nskiplen := 4
+	nretlen := 4
+	qlen := len(query)
+	msglen := MSGHEADER_LEN + flagslen + collname_len + nskiplen + nretlen + qlen
+	out := make([]byte, msglen)
+	// msg header
+	binary.LittleEndian.PutUint32(out[0:], uint32(msglen))
+	binary.LittleEndian.PutUint32(out[12:], OP_QUERY)
+	// query msg
+	idx := MSGHEADER_LEN + flagslen
+	copy(out[idx:idx+collname_len], []byte(collname))
+	idx += collname_len + nskiplen
+	binary.LittleEndian.PutUint32(out[idx:idx+nretlen], 1)
+	idx += nretlen
+	copy(out[idx:idx+qlen], query)
+	return out
+}
+
+// getOpMsg returns a mongodb OP_MSG message containing the specified BSON-encoded command.
+// section expected to be BSON byte array.
+func getOpMsg(section []byte) ([]byte) {
+	flagslen := 4
+	slen := len(section)
+	msglen := MSGHEADER_LEN + flagslen + slen
+	out := make([]byte, msglen)
+	// msg header
+	binary.LittleEndian.PutUint32(out[0:], uint32(msglen))
+	binary.LittleEndian.PutUint32(out[12:], OP_MSG)
+	// command msg
+	idx := MSGHEADER_LEN + flagslen
+	copy(out[idx:idx+slen], []byte(section))
+	return out
+}
+
+// getBuildInfoOpMsg returns a mongodb "OP" message containing query to retrieve MongoDB build info.
+func getBuildInfoOpMsg() ([]byte) {
+	// gleaned from tshark
+	section_payload, err := bson.Marshal(bson.M{ "buildinfo": 1, "$db": "admin" })
+	if err != nil {
+		// programmer error
+		log.Fatal("Invalid BSON")
```

Comment:
It would be nice if the specific error got logged too -- e.g. `log.Fatalf("Invalid BSON: %v", err)` or `...%s", err.String())`

Reply:
Good call